core/schema/modulesource.go

```diff
@@ -41,27 +45,41 @@ func (s *moduleSchema) moduleSource(ctx context.Context, query *core.Query, args
 
 	switch src.Kind {
 	case core.ModuleSourceKindLocal:
-		if filepath.IsAbs(modPath) {
-			return nil, fmt.Errorf("local module source root path is absolute: %s", modPath)
+		if filepath.IsAbs(refWithoutVersion) {
+			return nil, fmt.Errorf("local module source root path is absolute: %s", refWithoutVersion)
 		}
+
+		// parseRefString fallbacked to as we did not manage to localKind
+		// log error in case we wrongfully parsed the remote ref
+		if err != nil {
+			logger := telemetry.GlobalLogger(ctx)
+			logger.Info("ref %s has not been parsed as a git remote: %v", args.RefString, err)
+		}
+
 		src.AsLocalSource = dagql.NonNull(&core.LocalModuleSource{
-			RootSubpath: modPath,
+			RootSubpath: refWithoutVersion,
 		})
 
 	case core.ModuleSourceKindGit:
-		if !isGitHub {
-			return nil, fmt.Errorf("for now, only github.com/ paths are supported: %q", args.RefString)
-		}
-
 		src.AsGitSource = dagql.NonNull(&core.GitModuleSource{})
 
-		segments := strings.SplitN(modPath, "/", 4)
-		if len(segments) < 3 {
-			return nil, fmt.Errorf("invalid github.com path: %s", modPath)
+		// resolve the user input ref without version to a git URL
+		userRefWithoutVersion, err := resolveGitURL(refWithoutVersion)
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve git ref %s: %w", refWithoutVersion, err)
 		}
 
-		src.AsGitSource.Value.URLParent = segments[0] + "/" + segments[1] + "/" + segments[2]
+		// resolve the extracted repo root to a git URL
+		computedRepoRoot, err := resolveGitURL(parsed.repoRoot.Repo)
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve git ref %s: %w", parsed.repoRoot.Repo, err)
+		}
+
+		// splits the root and subdir from the repo root path and the ref without version path
+		root, subdir := splitRootAndSubdir(refWithoutVersion, computedRepoRoot.Path, userRefWithoutVersion.Path)
+		root = computedRepoRoot.Host + "/" + root
```

Comment:
This logic seems *very* complicated, it's split up into a bunch of functions, I'm struggling to understand what's actually going on.

Why isn't this as simple as getting `computedRepoRoot.Host + "/" + computedRepoRoot.Path`? Why do we need to reference `refWithoutVersion` again?

Also, isn't `parsed.repoRoot.Root` equivalent to `subdir` at this point? If it's not, what's the difference, and why isn't this logic part of the `parseRefString` logic earlier?

Reply:
I refactored the code to move it to its own package (to be shared with cmd/flags package to unify the directory logic)